contractcourt/htlc_timeout_resolver.go

```diff
@@ -372,13 +432,83 @@ func (h *htlcTimeoutResolver) Resolve() (ContractResolver, error) {
 
 	var reports []*channeldb.ResolverReport
 
+	claimOutpoint := h.htlcResolution.ClaimOutpoint
+
+	// If the sweeper is handling the second level transaction, wait for
+	// the CSV lock to expire.
+	if h.htlcResolution.SignDetails != nil {
+		h.reportLock.Lock()
+		h.currentReport.Stage = 2
+		h.reportLock.Unlock()
+
+		waitHeight := uint32(spend.SpendingHeight) +
+			h.htlcResolution.CsvDelay - 1
+
+		log.Infof("%T(%x): waiting for CSV lock to expire at height %v",
+			h, h.htlc.RHash[:], waitHeight)
+
+		if err := waitForHeight(waitHeight, h.Notifier, h.quit); err != nil {
+			return nil, err
+		}
+
+		// Find output index of the second-level, as the sweeper might
+		// have re-arranged outputs.
+		txOut := h.htlcResolution.SignedTimeoutTx.TxOut[0]
+		index := -1
+		for i, o := range spend.SpendingTx.TxOut {
+			if bytes.Equal(o.PkScript, txOut.PkScript) {
+				index = i
+			}
+		}
+
+		if index == -1 {
+			err := fmt.Errorf("unable to find second-level "+
+				"output %v on spending tx %v", txOut,
+				spend.SpenderTxHash)
+			log.Error(err)
+			return nil, err
+		}
+
+		op := &wire.OutPoint{
+			Hash:  *spend.SpenderTxHash,
+			Index: uint32(index),
+		}
+
+		log.Infof("%T(%x): CSV lock expired, offering second-layer "+
+			"output to sweeper: %v", h, h.htlc.RHash[:], op)
+
+		inp := input.NewCsvInput(
+			op, input.HtlcOfferedTimeoutSecondLevel,
+			&h.htlcResolution.SweepSignDesc,
+			h.broadcastHeight,
+			h.htlcResolution.CsvDelay,
+		)
+
+		_, err = h.Sweeper.SweepInput(
+			inp,
+			sweep.Params{
+				Fee: sweep.FeePreference{
+					ConfTarget: sweepConfTarget,
+				},
+			},
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		// TODO(halseth): need to checkpoint?
```

Comment:
Either you need to checkpoint, then gate of that new state, or make sure all the logic can be replayed above regardless of which transactions has been broadcast/presented/confirmed.

Reply:
Checkpointed.